File: cpp/FuelFabrication.cpp
// Fuel Fabrication Class

#include "FuelFabrication.h"

/*** Common Function ***/

std::set<std::string> bright::make_fuel_fab_params_set(MassStreams * mss, std::set<std::string> orig_set)
{
    std::set<std::string> new_set (orig_set);

    for ( MassStreams::iterator ms=(*mss).begin() ; ms != (*mss).end(); ms++ )
    {
        new_set.insert( "Weight_" + (*ms).first );
        new_set.insert( "deltaR_" + (*ms).first );
    };

    return new_set;
};

/*****************************************************/
/*** FuelFabrication Component Class and Functions ***/
/*****************************************************/

FuelFabrication::FuelFabrication()
{
};

FuelFabrication::FuelFabrication(std::string n) : FCComp(n)
{
};

FuelFabrication::FuelFabrication(std::set<std::string> paramtrack, std::string n) : FCComp(paramtrack, n)
{
};

FuelFabrication::FuelFabrication(MassStreams mss, MassWeights mws_in, Reactor1G r, std::string n) : \
    FCComp(bright::make_fuel_fab_params_set(&mss), n)
{
    initialize(mss, mws_in, r);
};


FuelFabrication::FuelFabrication(MassStreams mss, MassWeights mws_in, Reactor1G r, std::set<std::string> paramtrack, std::string n) : \
    FCComp(bright::make_fuel_fab_params_set(&mss, paramtrack), n)
{
    initialize(mss, mws_in, r);
};


FuelFabrication::~FuelFabrication()
{
};



void FuelFabrication::initialize(MassStreams mss, MassWeights mws_in, Reactor1G r)
{
    /** Sets the fuel fabrication specific parameters.
     *  Must be done once at the beginning of fuel fabrication object life.
     */

    mass_streams = mss;
    mass_weights_in = mws_in;

    reactor = r;

    calc_deltaRs();
};



void FuelFabrication::calc_deltaRs()
{
    // Calculates the deltaR for each mass stream
    for (MassStreams::iterator mss = mass_streams.begin(); mss != mass_streams.end(); mss++)
    {
        MassStream ms = (*(*mss).second);
        ms.normalize();

        double dR = reactor.calc_deltaR(ms);

        deltaRs[(*mss).first] = dR;
        ms.mass = mass_weights_in[(*mss).first];
    };
};


MassStream FuelFabrication::calc_core_input()
{
    CompDict cd;
    MassStream core_input (cd, 0.0);

    for (MassWeights::iterator mws = mass_weights_out.begin(); mws != mass_weights_out.end(); mws++)
    {
        core_input = core_input + ( (*mass_streams[(*mws).first]) * (*mws).second );
    };    

    core_input.normalize();
    core_input.name = "CoreInput";
    return core_input;
};


void FuelFabrication::calc_mass_ratios()
{
    // Find the key mass stremas based on which two have weights that are less than zero
    std::string key_a;
    std::string key_b;

    bool found_a = false;

    for (MassWeights::iterator mws = mass_weights_in.begin(); mws != mass_weights_in.end(); mws++)
    {
        if ((*mws).second < 0.0)
        {
            if (!found_a)
            {
                key_a = (*mws).first;
                found_a = true;
            }
            else
            {
                key_b = (*mws).first;
                break;
            };
        };
    };

    // deltaR for key a
    MassStream ms_a = *mass_streams[key_a];
    ms_a.normalize();
    double dR_a = reactor.calc_deltaR( ms_a );

    // deltaR for key b
    MassStream ms_b = *mass_streams[key_b];
    ms_b.normalize();
    double dR_b = reactor.calc_deltaR( ms_b );

    //First Guess for key mass stream masses; each get half of the remaining mass space.
    double top_up_mass_space = 1.0; 
    for (MassWeights::iterator mws = mass_weights_in.begin(); mws != mass_weights_in.end(); mws++)
    {
        if (0.0 <= (*mws).second)
            top_up_mass_space = top_up_mass_space - (*mws).second;
    };


    // Initialize mass_weights_out as a copy of mass_weights_in
    for (MassWeights::iterator mws = mass_weights_in.begin(); mws != mass_weights_in.end(); mws++)
    {
        mass_weights_out[(*mws).first] = (*mws).second;
    };

    double dR_guess;
    MassStream core_input;

    double k_a, k_b;
    double sign_a, sign_b;


    // Find bound for All Mass Stream A
    mass_weights_out[key_a] = top_up_mass_space;
    mass_weights_out[key_b] = 0.0;
    core_input = calc_core_input();
    dR_guess = reactor.calc_deltaR( core_input );
    k_a = reactor.batchAveK( reactor.TargetBU );
    sign_a = (1.0 - k_a) / fabs(1.0 - k_a);

    // Find bound for All Mass Stream B
    mass_weights_out[key_a] = 0.0;
    mass_weights_out[key_b] = top_up_mass_space;
    core_input = calc_core_input();
    dR_guess = reactor.calc_deltaR( core_input );
    k_b = reactor.batchAveK( reactor.TargetBU );
    sign_b = (1.0 - k_b) / fabs(1.0 - k_b);

    // Ensure calculation is possible
    if (sign_a == sign_b)
        throw BadFuelWeights(k_a, k_b);

    //Continue nomrally
    mass_weights_out[key_a] = top_up_mass_space * 0.5;
    mass_weights_out[key_b] = top_up_mass_space - mass_weights_out[key_a];

    //Calculate delta R for the Guess
    core_input = calc_core_input();
    dR_guess = reactor.calc_deltaR( core_input );

    int n;
    double k;
    double dMass;
    
    k = reactor.batchAveK( reactor.TargetBU );
    n = 0;
    if (0 < FCComps::verbosity)
        std::cout << n << ") " << k << " "; 

    while (0.001 < fabs(1.0 - k) && n < 10)
    {
        //Adjust Masses based on pertubation guess.
        dMass = - dR_guess / (dR_a - dR_b);
        mass_weights_out[key_a] = mass_weights_out[key_a] + dMass;
        mass_weights_out[key_b] = mass_weights_out[key_b] - dMass;

        //Recalculate core parameters for new masses guess
        core_input = calc_core_input();
        dR_guess = reactor.calc_deltaR( core_input );
        k = reactor.batchAveK( reactor.TargetBU );
        n = n+1;
        if (0 < FCComps::verbosity)
            std::cout << k << " ";
    };

    if (0 < FCComps::verbosity)
        std::cout << "\n\n";
    
};


MassStream FuelFabrication::doCalc()
{
    calc_mass_ratios();
    ms_prod = calc_core_input();
    return ms_prod;
};


MassStream FuelFabrication::doCalc(MassStreams mss, MassWeights mws_in, Reactor1G r)
{
    initialize(mss, mws_in, r);
    return doCalc();
};


void FuelFabrication::setParams ()
{
    for ( MassWeights::iterator mws = mass_weights_in.begin(); mws != mass_weights_in.end(); mws++ )
    {
        std::string mw_key ("Weight_" + (*mws).first);
        std::string dR_key ("deltaR_" + (*mws).first);

        params_prior_calc[mw_key]  = mass_weights_in[(*mws).first];
        params_after_calc[mw_key] = mass_weights_out[(*mws).first];

        params_prior_calc[dR_key]  = deltaRs[(*mws).first];
        params_after_calc[dR_key] = deltaRs[(*mws).first];
    };
};

File: cpp/fuel_fabrication.py
class_ds = \
"""Fuel Fabrication Fuel Cycle Component Class.  Daughter of FCComp class.

Parameters
----------
mats : dict or map or None, optional 
    A dictionary whose keys are string labels (eg, "U-235", "TRU", "My Fuel") 
    and whose values are mass streams.  For example::

        materials = {
            "U235": Material({922350: 1.0}, 1.0, "U-235"),
            "U236": Material({922360: 1.0}, 1.0, "U-236"),
            "U238": Material({922380: 1.0}, 1.0, "U-238"),
            }

    would be valid for a light water reactor.
mws_in : dict or map or None, optional 
    A dictionary whose keys are the same as for materials and whose values are the 
    associated weight (float) for that stream.  If a material should be allowed to 
    vary (ie optimized over), specify its weight as a negative number. For instance::

        mass_weights_in = {
            "U235": -1.0,
            "U236": 0.005,
            "U238": -1.0,
            }

    would be valid for a light water reactor with half a percent of U-236 always 
    present.
r : Reactor1G or None, optional 
    An instance of a Reactor1G class to fabricate fuel for.
paramtrack : list of str or None, optional 
    Additional parameters to track, if any.    
n : str, optional 
    The name of the fuel fabrication fuel cycle component instance.

"""

desc = {
    'docstrings': {
        'module': "Python wrapper for fuel fabrication.",
        'class': class_ds,
        'attrs': {},
        'methods': {},
        },
    'attrs': {},
    'extra': {},
    }

desc['docstrings']['attrs']['materials'] = \
"A mapping of materials which are mixed to create a valid fuel form for reactor."

desc['docstrings']['attrs']['mass_weights_in'] = \
"""A mapping representing the initial specification of the mass weights.  
Exactly two of these should have negative values."""

desc['docstrings']['attrs']['mass_weights_out'] = \
"""A mapping representing the mass weights that are calculated to generate 
a valid fuel from the materials  provided."""

desc['docstrings']['attrs']['deltaRs'] = \
"""A mapping representing the deltaR values of each of the materials."""

desc['docstrings']['attrs']['reactor'] = \
"""An instance of a Reactor1G class that the mat_prod is valid as a fuel for."""

desc['docstrings']['methods']['initialize'] = \
"""The initialize() method takes the appropriate materials, input mass weights,
and a reactor and resets the current FuelFabrication instance.

Parameters
----------
mats : dict
    A dictionary whose keys are string labels and whose values are mass streams.  
mws_in : dict 
    A dictionary whose keys are the same as for materials and whose values are the 
    associated weight (float) for that stream.
r : Reactor1G 
    An instance of a Reactor1G class to fabricate fuel for.

"""

desc['docstrings']['methods']['calc_params'] = \
"""Here the parameters for FuelFabrication are set.  For example::

    self.params_prior_calc["Weight_U235"] = self.mass_weights_in["U235"]
    self.params_after_calc["Weight_U235"] = self.mass_weights_out["U235"]

    self.params_prior_calc["deltaR_U235"] = self.deltaRs["U235"]
    self.params_after_calc["deltaR_U235"] = self.deltaRs["U235"]

    self.params_prior_calc["Weight_U238"] = self.mass_weights_in["U238"]
    self.params_after_calc["Weight_U238"] = self.mass_weights_out["U238"]

    self.params_prior_calc["deltaR_U238"] = self.deltaRs["U238"]
    self.params_after_calc["deltaR_U238"] = self.deltaRs["U238"]

"""

desc['docstrings']['methods']['calc_deltaRs'] = \
"""Computes deltaRs for each mass stream."""


desc['docstrings']['methods']['calc_core_input'] = \
"""Computes the core input material that becomes mat_prod based on materials and 
mass_weights_out.

Returns
-------
core_input : Material 
    mat_prod.
"""

desc['docstrings']['methods']['calc_mass_ratios'] = \
"""Calculates mass_weights_out by varying the values of the two parameter which had 
negative values in mass_weights_in.  Therefore, this is the portion of the code that 
performs the optimization calculation.
"""

desc['docstrings']['methods']['calc'] = \
"""This method performs an optimization calculation on all input materials to 
determine the mass ratios that generate the correct fuel form for the reactor.
It then compiles the fuel and returns the resultant material. 

Parameters
----------
mats : dict or None, optional
    A dictionary whose keys are string labels and whose values are materials.  
mws_in : dict or None, optional
    A dictionary whose keys are the same as for materials and whose values are 
    the associated weight (float) for that material.
r : Reactor1G or None, optional 
    An instance of a Reactor1G class to fabricate fuel for.

Returns
-------
core_input : Material 
    mat_prod

"""
